[RISCV][GISel] Add ISel supports for SHXADD from Zba extension

mshockwave opened the pull request:
This patch consists of porting SDISel patterns of SHXADD instructions to GISel.
Note that `non_imm12`, a predicate that was implemented with `PatLeaf`, is now turned into a `PatFrag` of `<op>_with_non_imm12` where `op` is the operator that uses `the non_imm12` operand, as GISel doesn't have equivalence of `PatLeaf` at this moment.

I'll put ISel supports for SHXADD_UW in a separate patch since the current one is getting a little big.

mshockwave committed:
[RISCV][GISel] Add ISel supports for SHXADD from Zba extension

This patch constitue of porting (SDISel) patterns of SHXADD
instructions.
Note that `non_imm12`, a predicate that was implemented with `PatLeaf`,
is now turned into a ComplexPattern to facilitate code reusing on
patterns that use it between SDISel and GISel.

mshockwave commented on llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:
```diff
@@ -105,6 +114,127 @@ RISCVInstructionSelector::selectShiftMask(MachineOperand &Root) const {
+    if (Mask.isShiftedMask()) {
+      unsigned Leading = XLen - Mask.getActiveBits();
+      unsigned Trailing = Mask.countr_zero();
+      // Given (and (shl y, c2), mask) in which mask has no leading zeros and c3
```
I'm copying the comments from RISCVISelDAGToDAG.cpp to here for better readability.

mshockwave committed:
fixup! [RISCV][GISel] Add ISel supports for SHXADD from Zba extension

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:
```diff
@@ -105,6 +114,127 @@ RISCVInstructionSelector::selectShiftMask(MachineOperand &Root) const {
   return {{[=](MachineInstrBuilder &MIB) { MIB.add(Root); }}};
 }
 
+// This complex pattern actually serves as a perdicate that is effectively
```
predicate*

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:
```diff
@@ -235,10 +235,7 @@ def SimmShiftRightBy3XForm : SDNodeXForm<imm, [{
-  auto *C = dyn_cast<ConstantSDNode>(N);
-  return !C || !isInt<12>(C->getSExtValue());
-}]>;
+def non_imm12 : ComplexPattern<XLenVT, 1, "selectNonImm12", [], [], 0>;
```
Can you add `let GISelPredicateCode` here to do this without changing to a complex pattern?

topperc replied:
After some digging, I think the answer is no and I'm sad about it.

mshockwave replied:
Correct, using `GISelPredicateCode` was my first approach until I found that the `GlobalISelEmitter` TG backend doesn't pick that up for leaf nodes.

topperc replied:
Would add GISelPredicateCode to something like this work https://github.com/topperc/llvm-project/commit/01205c136d77f682063a1218790f7cf0b3760fdc

mshockwave replied:
> Would add GISelPredicateCode to something like this work [topperc@01205c1](https://github.com/topperc/llvm-project/commit/01205c136d77f682063a1218790f7cf0b3760fdc)

This works, though I'm a little concerned that this might create too many boilerplate code in the future, since there needs to be a Predicate TG record for every opcode that goes with non_imm12 (even we abstract the real predicate logics into a function or something). What do you think? 

Also, interestingly `GISelPredicateCode` doesn't dance well with `PredicateCodeUsesOperands`: it SEGFAULT llvm-tblgen in our case, despite the fact that there are tests for this exact combination (in `test/TableGen/GlobalISelEmitterCustomPredicate.td`).  I can't find an obvious fix for llvm-tblgen but writing a non-`PredicateCodeUsesOperands` predicate code also works.

topperc replied:
I used `PredicateCodeUsesOperands` so that I could know which operand wasn't the `shl` since add is commutable and tblgen will generate both patterns.

mshockwave replied:
> Would add GISelPredicateCode to something like this work [topperc@01205c1](https://github.com/topperc/llvm-project/commit/01205c136d77f682063a1218790f7cf0b3760fdc)

Done: it's no longer using ComplexPattern of `non_imm12` but PatFrag of `<op>_with_non_imm12` instead

mshockwave committed:
(Staging) Use GISelPredicateCode in all SHXADD patterns

But since there is a bug in llvm-tblgen that crashes itself whenever a
ComplexPattern failed to be imported with `PredicateUsesOperands` +
`GISelPredicateCode`, we preserve the original `non_imm12` (PatLeaf) and
leave all `SHXADD_UW` patterns untouched.

mshockwave committed:
fixup! (Staging) Use GISelPredicateCode in all SHXADD patterns

mshockwave committed:
fixup! (Staging) Use GISelPredicateCode in all SHXADD patterns

mshockwave commented:
Note that this patch depends on #68125 so the buildbot failures in this patch are expected.

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:
```diff
@@ -105,6 +112,110 @@ RISCVInstructionSelector::selectShiftMask(MachineOperand &Root) const {
+    LeftShift = false;
+
+  if (LeftShift.has_value())
+    if (Mask.isShiftedMask()) {
```
Can we merge this condition with the previous if?

topperc replied:
Was this comment addressed?

mshockwave replied:
It is addressed now

mshockwave commented:
ping

mshockwave committed:
fixup! (Staging) Use GISelPredicateCode in all SHXADD patterns

topperc reviewed (approved):
LGTM

mshockwave committed:
Merge branch 'main' into gisel-riscv-zba-shift

mshockwave committed:
fixup! (Staging) Use GISelPredicateCode in all SHXADD patterns

mshockwave committed:
fixup! (Staging) Use GISelPredicateCode in all SHXADD patterns
